src/daft-io/src/object_io.rs

```diff
@@ -112,10 +118,240 @@ pub(crate) trait ObjectSource: Sync + Send {
     }
 }
 
+pub(crate) async fn glob(
+    source: Arc<dyn ObjectSource>,
+    glob: &str,
+) -> super::Result<BoxStream<'static, super::Result<FileMetadata>>> {
+    // If no special characters, we fall back to ls behavior
+    let full_fragment = GlobFragment::new(glob);
+    if !full_fragment.has_special_character() {
+        let glob = full_fragment.escaped_str().to_string();
+        return Ok(stream! {
```

Comment:
I've been having a lot of trouble with [StreamExt::filter](https://docs.rs/futures-preview/latest/futures/stream/trait.StreamExt.html#method.filter) , and hence ended up wrapping it in the stream macro. For some reason it makes lifetimes/typing very unhappy.

```
source
                .iter_dir(glob.as_str(), Some("/"), None)
                .await?
                .filter(
                    filter_directories
                ).boxed()
```

This throws me an error on the `.boxed()` call:
```
error[E0599]: the method `boxed` exists for struct `Filter<Pin<Box<dyn Stream<Item = Result<FileMetadata, Error>> + Send>>, impl Future<Output = bool>, ...>`, but its trait bounds were not satisfied
   --> src/daft-io/src/object_io.rs:139:19
    |
134 | /             source
135 | |                 .iter_dir(glob.as_str(), Some("/"), None)
136 | |                 .await?
137 | |                 .filter(
138 | |                     filter_directories
139 | |                 ).boxed());
    | |                  -^^^^^ method cannot be called due to unsatisfied trait bounds
    | |__________________|
    | 
    |
   ::: /Users/jaychia/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/filter.rs:12:1
    |
12  | / pin_project! {
13  | |     /// Stream for the [`filter`](super::StreamExt::filter) method.
14  | |     #[must_use = "streams do nothing unless polled"]
15  | |     pub struct Filter<St, Fut, F>
...   |
24  | |     }
25  | | }
    | | -
    | | |
    | |_doesn't satisfy `_: StreamExt`
    |   doesn't satisfy `_: Stream`
    |
    = note: the full type name has been written to '/Users/jaychia/code/Daft/target/debug/deps/daft_io-b6c2e5c24007c5f0.long-type-17637036591917212395.txt'
    = note: the following trait bounds were not satisfied:
            `futures::stream::Filter<Pin<Box<dyn futures::Stream<Item = std::result::Result<FileMetadata, Error>> + std::marker::Send>>, impl futures::Future<Output = bool>, for<'a> fn(&'a std::result::Result<FileMetadata, Error>) -> impl futures::Future<Output = bool> {filter_directories}>: futures::Stream`
            which is required by `futures::stream::Filter<Pin<Box<dyn futures::Stream<Item = std::result::Result<FileMetadata, Error>> + std::marker::Send>>, impl futures::Future<Output = bool>, for<'a> fn(&'a std::result::Result<FileMetadata, Error>) -> impl futures::Future<Output = bool> {filter_directories}>: StreamExt`
            `&futures::stream::Filter<Pin<Box<dyn futures::Stream<Item = std::result::Result<FileMetadata, Error>> + std::marker::Send>>, impl futures::Future<Output = bool>, for<'a> fn(&'a std::result::Result<FileMetadata, Error>) -> impl futures::Future<Output = bool> {filter_directories}>: futures::Stream`
            which is required by `&futures::stream::Filter<Pin<Box<dyn futures::Stream<Item = std::result::Result<FileMetadata, Error>> + std::marker::Send>>, impl futures::Future<Output = bool>, for<'a> fn(&'a std::result::Result<FileMetadata, Error>) -> impl futures::Future<Output = bool> {filter_directories}>: StreamExt`
            `&mut futures::stream::Filter<Pin<Box<dyn futures::Stream<Item = std::result::Result<FileMetadata, Error>> + std::marker::Send>>, impl futures::Future<Output = bool>, for<'a> fn(&'a std::result::Result<FileMetadata, Error>) -> impl futures::Future<Output = bool> {filter_directories}>: futures::Stream`
            which is required by `&mut futures::stream::Filter<Pin<Box<dyn futures::Stream<Item = std::result::Result<FileMetadata, Error>> + std::marker::Send>>, impl futures::Future<Output = bool>, for<'a> fn(&'a std::result::Result<FileMetadata, Error>) -> impl futures::Future<Output = bool> {filter_directories}>: StreamExt`

```

Not sure what's happening :(

Reply:
I've been having some trouble with streams and lifetimes. The `stream!` macro seems to correctly move values into the Future so that it owns it? For instance:

```
return Ok(source.iter_dir(glob.as_str(), Some("/"), None).await?.boxed());
```

The above throws an error:

```
cannot return value referencing local variable `source`
returns a value referencing data owned by the current function
```

I'm guessing this is because `source` is on the local stack, and does not correctly move into the returned Future since the function signature here is: `async iter_dir(&self, ...)` and the returned Future holds a reference to the local stack's `source`.

Reply:
One workaround I found is to change the function signature of `iter_dir` to: `iter_dir(self: Arc<Self>, ...) -> BoxStream<'a , ...> where Self: 'a`

This forces the BoxStream to own an `Arc<dyn ObjectSource>` instead of `&dyn ObjectSource`, so it doesn't hold a reference to some local Stack variable.

This in turn lets us return BoxStreams that contain local `Arc<dyn ObjectSource>` objects. Otherwise, we will have to immediately consume them in the local block before the Arc goes out of scope.

Seems a little heavy-handed though? Wdyt?

Reply:
that's cause `stream!` is moving all the values into a closure whereas source in `source.iter_dir` is being coerced into a reference. You can do your own explict move closure instead.

Reply:
We can go through this tomorrow in person, I wasn't able to get lifetimes working correctly...

`return stream!{ let s = source.iter_dir(...); while let Some(_) = s.next().await{...} };` moves `source` into a wrapper stream, which then owns both the `source` and the resultant `BoxStream` from `source.iter_dir`, making this new stream legal to return.

Returning the stream directly however (`return source.iter_dir(...).filter(...).boxed();`) returns a stream that holds a reference to `source` which is on the local stack. Even if I wrap this in a move closure, `source` will still remain on the local stack.